File: Ruby/CodeWars/8kyu-RockPaperScissors.rb
=begin 
Rock Paper Scissors
Let's play! You have to return which player won! In case of a draw return Draw!.
Examples:
rps('scissors','paper') // Player 1 won!
rps('scissors','rock') // Player 2 won!
rps('paper','paper') // Draw!
=end

def rps(p1, p2)
  return 'Draw!' if player1 == player2
  wins = [['paper', 'rock'], ['rock', 'scissors'], ['scissors', 'paper']]
  wins.include?([player1, player2]) ? 'Player 1 Wins!' : 'Player 2 Wins!'
end

  puts rps('rock', 'scissors')
  # => "Player 1 won!"
  puts rps('scissors', 'paper')
  # => "Player 1 won!"
  puts rps('paper', 'rock')
  # => "Player 1 won!"
  puts rps('scissors', 'rock')
  # => "Player 2 won!"
  puts rps('paper', 'scissors')
  # => "Player 2 won!"
  puts rps('rock', 'paper')
  # => "Player 2 won!"
  puts rps('rock', 'rock')
  # => 'Draw!'
  puts rps('scissors', 'scissors')
  # => 'Draw!'
  puts rps('paper', 'paper')
  # => 'Draw!'

=begin ############################################# OTHERS SOLUTIONS 
2)
def rps(p1, p2)
  beatmap = { 'scissors' => 'paper', 'paper' => 'rock', 'rock' => 'scissors' }

  if p1 == p2
    "Draw!"
  elsif beatmap[p1] == p2
    "Player 1 won!"
  else
    "Player 2 won!"
  end
end
----------------------------------------------------------------------
3)
def rps(p1, p2)
  return 'Draw!' if p1 == p2
  
  result = win_matrix[p1.to_sym][p2.to_sym]
  
  "Player #{result} won!"
end

def win_matrix
  { 
    rock:     { paper: 2, scissors: 1 },
    paper:    { scissors: 2, rock: 1 },
    scissors: { rock: 2, paper: 1 }
  }
end
----------------------------------------------------------------------
4)
def rps(p1, p2)
  return 'Draw!' if p1 == p2
  results = [['scissors', 'paper'], ['paper', 'rock'], ['rock', 'scissors']]
  return results.include?([p1, p2]) ? 'Player 1 won!' : 'Player 2 won!'
end
----------------------------------------------------------------------
5)

----------------------------------------------------------------------
6)

----------------------------------------------------------------------
7)

----------------------------------------------------------------------
8)

----------------------------------------------------------------------
9)

----------------------------------------------------------------------
10)

=end

=begin THRASH CAN ****************************************************
=end

File: JS/CodeWars/Storage/8kyu-RockPaperScissors.js
/* 
Rock Paper Scissors!
rps('scissors','paper') // Player 1 won!
rps('scissors','rock') // Player 2 won!
rps('paper','paper') // Draw!
*/
const rps = (p1, p2) => {
   wins = { 'scissors' : 'paper', 'paper' : 'rock', 'rock' : 'scissors' }
   if( p1 == p2 ){return 'Draw!'}
   if(wins[p1]== p2){
      return "Player 1 won!";
   }
   else{
      return "Player 2 won!";
   }
};

console.log(rps('rock', 'scissors'));
// : getMsg(1)
console.log(rps('scissors', 'paper'));
// : getMsg(1)
console.log(rps('paper', 'rock'));
// : getMsg(1)
console.log(rps('scissors', 'rock'));
// : getMsg(2)
console.log(rps('paper', 'scissors'));
// : getMsg(2)
console.log(rps('rock', 'paper'));
// : getMsg(2)
console.log(rps('rock', 'rock'));
// : 'Draw!'
console.log(rps('scissors', 'scissors'));
// : 'Draw!'
console.log(rps('paper', 'paper'));
// : 'Draw!'

/* ############################################# OTHERS SOLUTIONS 
2)
const rps = (p1, p2) => {
  if(p1 === p2) {
    return 'Draw!'
  }; 
  return `Player ${/rockscissors|scissorspaper|paperrock/.test(p1+p2)? 1 : 2} won!`;
}
----------------------------------------------------------------------
3)

rps=(a,b)=>a==b?'Draw!':`Player ${/rp|ps|sr/.test(a[0]+b[0])?2:1} won!`;
----------------------------------------------------------------------
4)
const rps = (a, b) => a == b ? 'Draw!' : `Player ${/rp|ps|sr/.test(a[0] + b[0]) ? 2 : 1} won!`;
----------------------------------------------------------------------
5)

----------------------------------------------------------------------
6)

----------------------------------------------------------------------
7)

----------------------------------------------------------------------
8)

----------------------------------------------------------------------
9)

----------------------------------------------------------------------
10)

*/

/* THRASH CAN ****************************************************

*/